File: algos/llist.py

class Node:
    def __init__(self, val):
        self.val = val
        self.next = self.prev = None

def llcircmake(*values):
    h1,t1 = llmake(*values)
    if h1 and t1:
        t1.next = h1
        h1.prev = t1
    return h1,t1

def llmake(*values):
    h1 = t1 = None
    for v in values:
        node = Node(v)
        h1,t1 = llinsert(h1, t1, node, t1, doubly=True)
    return h1,t1

def listlen(node):
    """ Return the length of a linkedlist.  """
    i = 0
    while node:
        node = node.next
        i += 1
    return i


def listcmp(l1, l2, cmpfunc=None):
    """ Compares two lists and return -1, 0 or 1 depending on whether
    values of first list <, == or > than values in the second list.

    A custom comparater func can also be provided to which the values of 
    the two nodes can be passed for comparison.
    """
    if not cmpfunc:
        cmpfunc = lambda x,y: x - y

    while l1 and l2:
        diff = cmpfunc(l1.val, l2.val)
        if diff != 0: return diff
        l1 = l1.next
        l2 = l2.next
    if not l1 and not l2: return 0
    elif not l1: return -1
    else: return 1

def lltail(node):
    """ Find the last node in a linked list.
        Note that this also works if our list is circular!
    """
    prev = None
    curr = node
    while curr:
        prev = curr
        curr = curr.next
        if curr == node: break  # will happen if a circular list
    return prev

def llmin(node):
    """
    Find the node with the minimum value in the list.
    Note that this also works if our list is circular!
    """
    min = node
    curr = node
    while curr:
        if curr.val < min.val: min = curr
        curr = curr.next
        if curr == node: break # means we are circular
    return min

def llmid(node):
    """ Find the middle node of a linked (or double linked) list even if the list is circular.
    middle index is len(list) // 2
    """
    prev = None
    slow = fast = node

    while slow and fast:
        # Moving fast first means we dont do any "advance" jumps
        fast = fast.next
        if fast in (node, None): break # 'node' here also checks for circular lists
        else:
            fast = fast.next

        prev = slow
        slow = slow.next

        # Check here if fast is ended otherwise dont bother going back again
        # Especially useful for circular lists
        if fast in (node, None): break
    return prev, slow
    
def llfind(head, findfunc):
    """ Given a head node, finds the first node that matches the criteria from the findfunc.

    Findfunc:
        takes 2 parameters: the prevnode and the current node being tested
        returns the node which maches the criteria along with the previous node (used for insertion)
        If no such node matches than None,None are returned.

    This method will work for circular lists too.
    """
    prev = None
    curr = head
    while curr:
        if findfunc(prev, curr):
            # found the node before which we have to insert
            break
        prev = curr
        curr = curr.next
        if curr == head: 
            # we are back to head after hitting the circular list
            # caller should check if curr == head
            break
    return prev, curr

def llinsert(head, tail, newnode, after=None, doubly=False):
    """
    Insert a newnode after a given 'after' node.  Our list is denoted by the head
    and tail nodes.  Head and/or tail can be null so this method takes care of
    returning the new head/tail after the insertion is complete.  Also all next
    (and prev if doubly linked list) pointers are updated and returned.

    If the after node is null then the item is prepended to the list (and a
    new head will be returned).

    NOTE - the node must *not* be in this list
    """
    if not head:
        return newnode, newnode

    newnode.next = None
    if doubly: newnode.prev = None

    if not after:
        # Prepend
        newnode.next = head
        if doubly: head.prev = newnode
        return newnode, tail

    # Save prev/next pointers
    prev = after
    next = after.next

    newnode.next = next

    if prev: prev.next = newnode
    if doubly:
        newnode.prev = prev
        if next: next.prev = newnode

    if after == tail:
        tail = newnode

    return head, tail

import unittest
class TestMethods(unittest.TestCase):
    def test_listlen(self):
        self.assertEqual(listlen(None), 0)

        h,t = llmake() ; listlen(h)
        self.assertEqual(listlen(h), 0)

        h,t = llmake(1,2,3,4,5) ; listlen(h)
        self.assertEqual(listlen(h), 5)

    def test_listcmp(self):
        h1,t1 = llmake(1,2,3,4,5)
        h2,t2 = llmake(1,2,3,4,5)
        self.assertEqual(listcmp(h1, h2), 0)

        h1,t1 = llmake(1,2,3,4,5)
        h2,t2 = llmake(1,2,3,4,5,6)
        self.assertEqual(listcmp(h1, h2), -1)

        h1,t1 = llmake(1,2,3,4,5)
        h2,t2 = llmake(1,2,3,4)
        self.assertEqual(listcmp(h1, h2), 1)

        h1,t1 = llmake(1,2,3,4,5)
        h2,t2 = llmake(1,2,3,4,7)
        listcmp(h1, h2)
        self.assertEqual(listcmp(h1, h2), -2)

    def test_lltail(self):
        h1,t1 = llmake(1,2,3,4,5)
        self.assertEqual(lltail(h1).val, 5)

    def test_llmin(self):
        h1,t1 = llmake(10, 4, 9, 11, 1, 200)
        self.assertEqual(llmin(h1).val, 1)
    
    def test_llmid(self):
        h1,t1 = llmake(1)
        prev,mid = llmid(h1)
        self.assertEqual((prev, mid.val), (None, 1))

        h1,t1 = llmake(1, 2)
        prev,mid = llmid(h1)
        self.assertEqual((prev.val, mid.val), (1, 2))

        h1,t1 = llmake(10, 4, 9, 11, 1, 200)
        prev,mid = llmid(h1)
        self.assertEqual((prev.val, mid.val), (9, 11))

        h1,t1 = llmake(10, 4, 9, 11, 1, 200, 300)
        prev,mid = llmid(h1)
        self.assertEqual((prev.val, mid.val), (9, 11))

    def test_llmid_circular(self):
        h1,t1 = llcircmake(1)
        prev,mid = llmid(h1)
        self.assertEqual((prev, mid.val), (None, 1))

        h1,t1 = llcircmake(1, 2)
        prev,mid = llmid(h1)
        self.assertEqual((prev.val, mid.val), (1, 2))

        h1,t1 = llcircmake(10, 4, 9, 11, 1, 200)
        prev,mid = llmid(h1)
        self.assertEqual((prev.val, mid.val), (9, 11))

        h1,t1 = llcircmake(10, 4, 9, 11, 1, 200, 300)
        prev,mid = llmid(h1)
        self.assertEqual((prev.val, mid.val), (9, 11))

if __name__ == '__main__':
    # Run with python -m unittest *.py
    unittest.main()

File: go/algos/ds/llist.go
// Single and Double LinkedList implementations
package ds

import "log"

type ListNode[T any] struct {
	Prev *ListNode[T]
	Val  T
	Next *ListNode[T]
}

// Creates a new list with a list of values
func NewListNode[T any](value T) (out *ListNode[T]) {
	return &ListNode[T]{Val: value}
}

func NewList[T any](doubly bool, values ...T) (head, tail *ListNode[T]) {
	for _, v := range values {
		newnode := &ListNode[T]{Val: v}
		head, tail = LNodeInsert(head, tail, newnode, tail, doubly)
	}
	return
}

// Return all values in a list
func (n *ListNode[T]) Values() (out []T) {
	for n != nil {
		out = append(out, n.Val)
		n = n.Next
	}
	return
}

func (n *ListNode[T]) FirstInvalidNode() int {
	var prev *ListNode[T]
	curr := n
	i := 0
	for curr != nil {
		if prev != nil {
			if prev.Next != curr {
				return i
			}
		}
		if curr.Prev != prev {
			return i
		}
		prev = curr
		curr = curr.Next
		i += 1
	}
	return -1
}

// Inserts a node after a given 'after' node.  Our list is denoted by the head and tail nodes.
// Head and/or tail can be null so this method takes care of returning the new head/tail after the
// insertion is complete.  Also all next (and prev if doubly linked) pointers are updated and
// returned.
//
// If the after node is null then the item is prepended to the list (and a new head will
// be returned).
//
// NOTE - newnode must NOT be in this list.  newnode's next and prev are first
// clobbered to nil so if newnode belongs to this, remove it first (with Delete)
func LNodeInsert[T any](head, tail, newnode *ListNode[T], after *ListNode[T], doubly bool) (*ListNode[T], *ListNode[T]) {
	// Ensure that newnode does not have a successor or a predecessor
	if newnode.Next != nil || newnode.Prev != nil {
		log.Println("Next or Prev is not nil....")
	}
	newnode.Next = nil
	if doubly {
		newnode.Prev = nil
	}

	// Empty list
	if head == nil {
		return newnode, newnode
	}

	// Insert at the front
	if after == nil {
		newnode.Next = head
		if doubly {
			head.Prev = newnode
		}
		return newnode, tail
	}

	// Save next/prev pointers
	newnode.Next = after.Next
	after.Next = newnode
	if doubly {
		if newnode.Next != nil {
			newnode.Next.Prev = newnode
		}
		newnode.Prev = after
	}

	if after == tail {
		tail = newnode
	}
	return head, tail
}

// Deletes a particular node.
// Parameter 'prev' is only required if we are talking about singly linked lists.
func LNodeDelete[T any](head, tail, node, prev *ListNode[T], doubly bool) (*ListNode[T], *ListNode[T]) {
	if head == nil {
		return nil, nil
	}

	// Replacing the head - this also takes care of when head == tail
	if node == head {
		next := head.Next

		// Clean up head next/prev
		head.Next = nil

		if next != nil {
			if doubly {
				next.Prev = nil
			}
			return next, tail
		} else {
			return nil, nil
		}
	}

	// Normal case where prev exists
	if doubly {
		prev = node.Prev
	} else if prev == nil || prev.Next != node {
		panic("In a singly linked list prev must be non nil and prev.next == node")
	}

	next := node.Next
	prev.Next = next
	if doubly {
		next.Prev = prev
	}

	if node == tail {
		tail = prev
	}
	return head, tail
}

// Starting at head of a singly linked list, returns the "previous" node for a given node.
// Not really needed for a doubly linked list.
func LNodePrev[T any](head, node *ListNode[T]) (prev *ListNode[T]) {
	if head == nil {
		return nil
	}

	// we are a singly LL, so find prev
	curr := head
	prev = nil
	for curr != nil {
		if curr == node {
			return prev
		}
		prev = curr
		curr = curr.Next
	}
	return nil
}

// cancatenates two linked lists
func LNodeCat[T any](h1, t1, h2, t2 *ListNode[T], doubly bool) (nh, nt *ListNode[T]) {
	if h1 == nil {
		return h2, t2
	}
	if h2 == nil {
		return h1, t1
	}
	t1.Next = h2
	if doubly {
		h2.Prev = t1
	}
	return h1, t2
}

// Compares the list starting at this node with another list starting at 'another' node
// and returns -1, 0 or 1 depending on the comparator func which compares the two values
func (n *ListNode[T]) Compare(another *ListNode[T], cmpfn func(a, b T) int) int {
	l1 := n
	l2 := another
	for l1 != nil && l2 != nil {
		v := cmpfn(l1.Val, l2.Val)
		if v != 0 {
			return v
		}
		l1 = l1.Next
		l2 = l2.Next
	}
	if l1 == nil && l2 == nil {
		return 0
	} else if l1 == nil {
		return -1
	} else {
		return 1
	}
}

// Finds the length of the list starting from this node.
func (n *ListNode[T]) Length() (l int) {
	for n != nil {
		l += 1
		n = n.Next
	}
	return
}

// Starting at a head node, finds the first node that matches the given matcher function.
// matcher:
// Takes 2 arguments: the prevnode and the currnode being tested and returns true if a match is found.
//
// Returns the node which matches the criteria along with the previous node.
// The previous node is useful for singly linked list incase an insertion has to
// happen at this point by the caller.
func (n *ListNode[T]) Find(matcher func(val T) bool) (prev *ListNode[T], curr *ListNode[T]) {
	curr = n
	for curr != nil {
		if matcher(curr.Val) {
			return prev, curr
		}
		prev = curr
		curr = curr.Next
	}
	return
}

// Find the tail starting from a given node.
// Even works for circular linked lists
func (n *ListNode[T]) Tail() *ListNode[T] {
	prev := n.Prev
	curr := n
	for curr != nil {
		prev = curr
		curr = curr.Next
		if curr == n {
			break
		}
	}
	return prev
}

// Find the node with the minimum value in the list.
// Note that this also works if our list is circular!
func (n *ListNode[T]) Min(less func(a, b T) bool) (prev *ListNode[T], min *ListNode[T]) {
	curr := n
	min = n
	p := n.Prev
	for curr != nil {
		if less(curr.Val, min.Val) {
			min = curr
			prev = p
		}
		p = curr
		curr = curr.Next
		if curr == n {
			break
		}
	}
	return
}

// Finds the midpoint node starting from this node (as well as its predecessor node).
// Note that this also works if our list is circular!
func (n *ListNode[T]) Mid() (prev *ListNode[T], mid *ListNode[T]) {
	slow := n
	fast := n
	for slow != nil && fast != nil {
		fast = fast.Next
		if fast == nil || fast == n {
			// Also checking for circular lists
			break
		} else {
			fast = fast.Next
		}

		prev = slow
		slow = slow.Next

		// Check here if fast ended otherwise dont bother going back again
		// Especially useful for circular lists
		if fast == n || fast == nil {
			break
		}
	}
	return prev, slow
}
